File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Random;
import java.util.stream.IntStream;

class DnDCharacter {

    static private final Random r = new Random();

    private int strength;
    private int dexterity;
    private int constitution;
    private int intelligence;
    private int wisdom;
    private int charisma;
    private int hitpoints;

    DnDCharacter() {
        strength = ability();
        dexterity = ability();
        constitution = ability();
        intelligence = ability();
        wisdom = ability();
        charisma = ability();
        hitpoints = 10 + modifier(constitution);
    }

    int ability() {
        IntStream dice = r.ints(4, 1, 7);
        dice = dice.sorted();
        dice = dice.skip(1);
        return dice.sum();
    }

    int modifier(int input) { return Math.floorDiv(input - 10, 2); }

    // Getters
    int getStrength() { return strength; }
    int getDexterity() { return dexterity; }
    int getConstitution() { return constitution; }
    int getIntelligence() { return intelligence; }
    int getWisdom() { return wisdom; }
    int getCharisma() { return charisma; }
    int getHitpoints() { return hitpoints; }
}

File: python/dnd-character/dnd_character.py
import secrets


def modifier(constitution):
    return (constitution - 10) // 2

class Character:
    def __init__(self):
        self.strength = Character.ability()
        self.dexterity = Character.ability()
        self.constitution = Character.ability()
        self.intelligence = Character.ability()
        self.wisdom = Character.ability()
        self.charisma = Character.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    @staticmethod
    def ability():
        four_random_rolls = [1 + secrets.randbelow(6) for _ in range(4)]
        return sum(sorted(four_random_rolls)[1:])